### ISSUE TITLE
Detect HTTP/2

### ISSUE THREAD

**Issue by Kriechi** [joushou/serve2#2]
Is it possible to detect HTTP/2?
The protocol is binary - but the connection preface should be very easy to spot:
https://tools.ietf.org/html/rfc7540#section-3.5

**Comment by joushou** [comment]
HTTP/2.0 is split into two, HTTP/2.0 with TLS (called "h2"), and HTTP/2.0 without TLS (called "h2c").

h2 is negotiated using TLS's protocol negotiation features, which I want to expose with my "hints" feature. This allows for transports to provide data for later protocol detection and handling. The "hints" branch implements this, although the design is still a bit WIP. Handing it to the Go HTTP server, however, is a bit more tricky, as it likes doing the TLS magic itself. There's an issue open, and I might be able to trick my way around this with a bit more work. Redirecting to elsewhere is no issue, though.

h2c, on the other hand, starts with a HTTP/1.1 request with an Upgrade header, where a HTTP/2.0 capable server will initiate and reply directly in 2.0 mode, with non-2.0 capable servers just ignoring the Upgrade request and replying as usual. This kind of request is a bit hard to distinguish from a regular HTTP/1.1 request, unless one reads the entire request, rather than just the first few bytes. This is a possibility, but in this case, I think the "correct" handling is just to detect HTTP/1.1, and dispatch it to a HTTP/2.0 capable server handling the upgrade.

h2c (*not* h2) can be initiated directly by using the aforementioned preamble, in case the client has received other knowledge regarding the servers support. This is, however, unlikely to be a common scenario, first of all due to most clients only implemented h2, and second of all due to the spec saying that one should not assume that a server previously supporting h2c would do it again, reducing the chances for using the preamble quite a lot.

TL;DR:
* h2 support is coming very soon through transport hints, although Go http server interaction is slightly tricky.
* h2c is just a matter of detecting HTTP/1.1 *or* the preamble, and sending it to a h2c capable server.

**Comment by Kriechi** [comment]
Great - thanks for the full explanation :+1: 
As far as I know, Firefox, Chrome, and IE/Edge only support `h2`,
so I think this is the most important feature.

**Comment by joushou** [comment]
I have this in the cleanup branch (as the cleanup branch spawned from the tlscheck branch... That's what happens when something suddenly bugs you!), and once it has been running functionally on my server for a bit, I'll merge that branch to master.

serve2d's experimental branch uses this feature to expose a "tlsmatcher", which right now allows you to make rules like proxy, including a dialTLS option which is needed to call a h2 server. Server name and negotiated protocol is maintained when using dialTLS. More values to check on will be added later, such as certificate checks.

This also enables serve2d to be an "SNI multiplexer", but just having a tls rule as well as a bunch of tlsmatcher rules for different server names.

**Comment by joushou** [comment]
TLSMatcher merged to master, which due to being able to check the negotiated protocol (NextProtos stuff), is able to detect HTTP/2.0 without knowing anything about that protocol in particular. With that, I'm closing this issue.
